src/leaflike/bookmarks.clj

```diff
@@ -16,98 +16,130 @@
             [clojure.string :as string]
             [ring.util.response :as res]))
 
-(defn parse-tags [tags]
-  (let [[tag-type tag-value] (s/conform ::bm-spec/tags tags)]
-    (case tag-type
-      :nil             []
-      :string          [tag-value]
-      :coll-of-strings tag-value)))
+(defn parse-input [f input]
+  (let [[input-type input-value] (s/conform f input)]
+    (case input-type
+      :nil []
+      :string [input-value]
+      :coll-of-strings input-value)))
+
+(defn associate-bookmark-with-users
+  [users bookmark-id]
+  (-> (user-db/get-user-ids-by-username users)
+      (utils/vals-from-list-of-id-maps)
+      (bm-db/bookmark-user bookmark-id)))
 
 (defn create
   [{:keys [params] :as request}]
   (if (s/valid? ::bm-spec/bookmark params)
     (let [user (hutils/get-user request)
           bookmark (-> (select-keys params [:title :url])
-                       (assoc :user_id (:id user)
+                       (assoc :created_by (:id user)
                               :created_at (utils/get-timestamp)))
-          tags (parse-tags (:tags params))
+          tags (parse-input ::bm-spec/tags (:tags params))
           bookmark-id (bm-db/create bookmark)
-          next-url (:next params "/bookmarks")]
+          next-url (:next params "/bookmarks")
+          username (get-in request [:session :username])
+          users (concat (parse-input ::bm-spec/collaborators (:collaborators params)) [username])]
+      (associate-bookmark-with-users users bookmark-id)
       (when (not-empty tags)
         (tags-db/create tags)
         (bm-db/tag-bookmark bookmark-id tags))
       (-> (res/redirect next-url)
           (assoc-in [:headers "Content-Type"] "text/html")
           (assoc-in [:flash :success-msg] "Bookmark added.")))
     (assoc (res/redirect "/bookmarks/add")
-           :flash {:error-msg (bm-spec/describe-errors params)})))
+      :flash {:error-msg (bm-spec/describe-errors params)})))
 
 (defn edit
   [{:keys [params] :as request}]
   (if (s/valid? ::bm-spec/bookmark params)
-    (let [user         (hutils/get-user request)
-          bookmark-id  (Integer/parseInt (:id params))
+    (let [user (hutils/get-user request)
+          bookmark-id (Integer/parseInt (:id params))
           updated-keys (select-keys params [:title :url])
-          tags         (parse-tags (:tags params))]
-      (bm-db/update-bookmark bookmark-id (:id user) updated-keys)
+          tags (parse-input ::bm-spec/tags (:tags params))
+          collaborators (parse-input ::bm-spec/collaborators (:collaborators params))]
+      (bm-db/update-bookmark bookmark-id updated-keys)
       (bm-db/remove-all-tags bookmark-id)
+      (bm-db/remove-all-bookmark-collaborators bookmark-id)
+      (when (not-empty collaborators)
+        (associate-bookmark-with-users collaborators bookmark-id))
```

Comment:
is this the best way to do this? remove all collaborators and then add again?

Reply:
@nid90 this is definitely not the best way to do this. I was most skeptical about this piece of code in this entire feature. I'm glad you brought it up. I initially wanted to send a diff from the frontend (a map containing removed and newly-added). I also tried writing a combination of insert and delete queries, but upon translating it into honeysql it got really messy and I kept running into problems. Honesysql's `update` lacks example and is also not very well documented. I wanted to ask about this specifically at the time of review. Can we discuss this in person? I'm not very happy with what's going on here and would like to get it right.